app/code/Magento/Sales/Model/Order/Shipment.php

```diff
@@ -435,19 +463,28 @@ public function addComment($comment, $notify = false, $visibleOnFront = false)
      */
     public function getCommentsCollection($reload = false)
     {
-        if (!$this->hasData(ShipmentInterface::COMMENTS) || $reload) {
+        if (!$this->hasData(self::COMMENTS_COLLECTION) || $reload) {
             $comments = $this->_commentCollectionFactory->create()
                 ->setShipmentFilter($this->getId())
                 ->setCreatedAtOrder();
-            $this->setComments($comments);
+            $this->setCommentsCollection($comments);
 
             if ($this->getId()) {
-                foreach ($this->getComments() as $comment) {
+                foreach ($this->getCommentsCollection() as $comment) {
                     $comment->setShipment($this);
                 }
             }
         }
-        return $this->getComments();
+        return $this->getData(self::COMMENTS_COLLECTION);
```

Comment:
You don't need to store `collection` in the `data` property, you could introduce own private property for the collection.

Reply:
Also, these changes break existing functionality. The `\Magento\Sales\Model\Order\Shipment::addComment` won't work anymore because `data[ShipmentInterface::COMMENTS]` doesn't contain `Comment` entities.

Reply:
addComment would continue to work as it worked before because it is using 
```
$this->getCommentsCollection()->addItem($comment);
``` 
To add new comment.

I don't agree that it should be stored as variable as this class extends DataObject for data storage purpose. For example, you would lose the ability to use setData, getData to change comments collection (what was probably intended by code that was there before I made my changes)

Reply:
[Relation::processRelation](https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/Sales/Model/ResourceModel/Order/Shipment/Relation.php#L71-L76) is responsible to persist shipment comment. After your changes, `$object->getComments()` will return an empty array and comments won't be saved. It's easy to check, try to create the shipment with comment from the Admin panel on the branch with your fix.

> I don't agree that it should be stored as variable as this class extends DataObject for data storage purpose. For example, you would lose the ability to use setData, getData to change comments collection

You shouldn't have a possibility to change the collection directly. To add new comments the `\Magento\Sales\Model\Order\Shipment::addComment` should be used.